src/storage/src/table/state_table.rs

```diff
@@ -113,171 +116,109 @@ impl<S: StateStore> StateTable<S> {
         Ok(())
     }
 
-    pub async fn iter(&self, epoch: u64) -> StorageResult<StateTableRowIter<'_, S>> {
-        let mem_table_iter = self.mem_table.buffer.iter().peekable();
-        StateTableRowIter::new(
+    pub async fn iter(
+        &self,
+        epoch: u64,
+    ) -> StorageResult<Pin<Box<impl Stream<Item = StorageResult<Option<Row>>> + Send + '_>>> {
+        let mem_table_iter = self.mem_table.buffer.iter();
+        Ok(Box::pin(StateTableRowIter::into_stream(
             &self.keyspace,
             self.column_descs.clone(),
             mem_table_iter,
             &self.order_types,
             epoch,
-        )
-        .await
+        )))
     }
 }
 
-/// `StateTableRowIter` is able to read the just written data (uncommited data).
-/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`
-pub struct StateTableRowIter<'a, S: StateStore> {
-    mem_table_iter: Peekable<MemTableIter<'a>>,
-    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
-
-    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
-    /// discarded.
-    cell_based_item: Option<(Vec<u8>, Row)>,
-    pk_serializer: OrderedRowSerializer,
+struct StateTableRowIter<S: StateStore> {
+    _phamtom: PhantomData<S>,
 }
-type MemTableIter<'a> = btree_map::Iter<'a, Row, RowOp>;
 
-enum NextOutcome {
-    MemTable,
-    Storage,
-    Both,
-    End,
-}
-impl<'a, S: StateStore> StateTableRowIter<'a, S> {
-    async fn new(
-        keyspace: &Keyspace<S>,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`.
+impl<S: StateStore> StateTableRowIter<S> {
+    /// This function scans kv pairs from the `shared_storage`(`cell_based_table`) and
+    /// memory(`mem_table`). If a record exist in both `cell_based_table` and `mem_table`, result
+    /// `mem_table` is returned according to the operation(RowOp) on it.
+    #[try_stream(ok = Option<Row>, error = StorageError)]
+    async fn into_stream<'a>(
```

Comment:
The weird thing here is that without `'a`, it errors.

Reply:
I think that's a compiler bug of the current generator, as in GAT async trait 🤣